source/language/pulses.rst

```diff
@@ -92,3 +94,30 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
 corresponding ``gate`` definitions. However, if a user provides a ``defcal`` for a symbol
 without a corresponding ``gate``, then we treat such operations like the ``opaque`` gates
 of prior versions of OpenQASM.
+
+Restrictions on defcal bodies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The contents of ``defcal`` bodies are subject to some restrictions. This is
+similar to how boxed expressions work.
+
+- They must have a definite length, regardless of the parameters passed in or
+  the state of the system when called. This allows the compiler to properly
+  resolve ``lengthof(...)`` calls and allowing for optimizations.
+- No control flow is allowed within a ``defcal`` block (with one exception,
+  see below). This follows from the definite length requirement.
+
+The sole exception to the second rule is a ``reset`` gate. The ``defcal`` for a
+``reset`` gate is permitted to have a single if statement, provided each branch
+of the if statement has definite and equivalent length.
```

Comment:
To play devil's advocate, why is it just restricted to a `reset` gate? Why not allow for any `defcal` block to have control-flow provided each branch has definite and equivalent length. Taking this one step further, for loops that have a compile-time resolvable `lengthof(...)` could also qualify. If the compiler is already resolving branches for a `reset`,  I don't see why it couldn't do the same for other control flow constructs.  I'm not sure if it would resolve @blakejohnson's `kludge` concern, but it captures the ultimate intention of definite length.

Reply:
Agreed. We started discussing this particular issue over in #102. In an internal discussion last week I think we came up with a workable alternative that @levbishop was going to write down. But the gist was allowing `defcal` or `defchannel` for anything with a compile-time computable duration.